File: python/src/wayid_verify/did.py
"""WayID DID validation + normalization.

Format: ``wayid:agent:{24-char-base58}``. Mirrors way-id ``src/lib/did/generator.ts``.
"""

import re
from typing import Optional

_WAYID_RE = re.compile(r"^wayid:agent:[1-9A-HJ-NP-Za-km-z]{24}$")
_BARE_RE = re.compile(r"^[1-9A-HJ-NP-Za-km-z]{24}$")

_PREFIX = "wayid:agent:"


def is_valid_wayid(wayid: str) -> bool:
    return bool(_WAYID_RE.match(wayid))


def normalize_wayid(value: str) -> Optional[str]:
    """Return the full DID for a full DID or bare 24-char tail, else ``None``."""
    if value is None:
        return None
    v = value.strip()
    if _WAYID_RE.match(v):
        return v
    if _BARE_RE.match(v):
        return _PREFIX + v
    return None


def bare_id(did: str) -> str:
    """Return the bare 24-char tail of a (normalized) DID."""
    return did[len(_PREFIX):] if did.startswith(_PREFIX) else did

File: js/src/did.ts
/**
 * WayID DID validation + normalization.
 *
 * Format: `wayid:agent:{24-char-base58}`. Copied (intentionally, to keep this
 * package dependency-free and standalone) from way-id `src/lib/did/generator.ts`.
 */

const WAYID_REGEX = /^wayid:agent:[1-9A-HJ-NP-Za-km-z]{24}$/;
const BARE_ID_REGEX = /^[1-9A-HJ-NP-Za-km-z]{24}$/;

export function isValidWayId(wayid: string): boolean {
  return WAYID_REGEX.test(wayid);
}

/**
 * Normalizes input into a full WayID DID. Accepts the canonical
 * `wayid:agent:<24 base58>` form or the bare 24-char base58 tail (issue #330).
 * Returns `null` if the input is neither shape.
 */
export function normalizeWayId(input: string): string | null {
  const trimmed = input.trim();
  if (WAYID_REGEX.test(trimmed)) return trimmed;
  if (BARE_ID_REGEX.test(trimmed)) return `wayid:agent:${trimmed}`;
  return null;
}

/** Returns the bare 24-char tail of a (normalized) DID. */
export function bareId(did: string): string {
  return did.startsWith("wayid:agent:")
    ? did.slice("wayid:agent:".length)
    : did;
}
